File: python/problems/leetcode/0019_remove_nth_node_from_end_of_list/solution.py
"""LeetCode 19: Remove Nth Node From End of List."""

from __future__ import annotations

from typing import Optional


class ListNode:
    def __init__(self, val: int = 0, next: Optional["ListNode"] = None) -> None:
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(
        self, head: Optional[ListNode], n: int
    ) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        fast: Optional[ListNode] = dummy
        slow: ListNode = dummy
        for _ in range(n + 1):
            assert fast is not None
            fast = fast.next
        while fast is not None:
            fast = fast.next
            assert slow.next is not None
            slow = slow.next
        assert slow.next is not None
        slow.next = slow.next.next
        return dummy.next


def to_list(head: Optional[ListNode]) -> list[int]:
    out: list[int] = []
    while head is not None:
        out.append(head.val)
        head = head.next
    return out


def from_list(vs: list[int]) -> Optional[ListNode]:
    dummy = ListNode()
    cur = dummy
    for v in vs:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next


if __name__ == "__main__":
    assert to_list(Solution().removeNthFromEnd(from_list([1, 2, 3, 4, 5]), 2)) == [
        1, 2, 3, 5,
    ]
    assert to_list(Solution().removeNthFromEnd(from_list([1]), 1)) == []
    assert to_list(Solution().removeNthFromEnd(from_list([1, 2]), 1)) == [1]
    print("0019 OK")

File: cpp/problems/leetcode/0019_remove_nth_node_from_end_of_list/solution.cpp
// LeetCode 19
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* n) : val(x), next(n) {}
};

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode dummy(0, head);
        ListNode* fast = &dummy;
        ListNode* slow = &dummy;
        for (int i = 0; i < n + 1; ++i) fast = fast->next;
        while (fast) {
            fast = fast->next;
            slow = slow->next;
        }
        ListNode* del = slow->next;
        slow->next = del->next;
        delete del;
        return dummy.next;
    }
};

static ListNode* fromList(vector<int> vs) {
    ListNode dummy;
    ListNode* cur = &dummy;
    for (int v : vs) {
        cur->next = new ListNode(v);
        cur = cur->next;
    }
    return dummy.next;
}

static vector<int> toList(ListNode* h) {
    vector<int> out;
    while (h) {
        out.push_back(h->val);
        h = h->next;
    }
    return out;
}

int main() {
    assert((toList(Solution().removeNthFromEnd(fromList({1, 2, 3, 4, 5}), 2)) ==
            vector<int>{1, 2, 3, 5}));
    assert((toList(Solution().removeNthFromEnd(fromList({1}), 1)) == vector<int>{}));
    assert((toList(Solution().removeNthFromEnd(fromList({1, 2}), 1)) == vector<int>{1}));
    cout << "0019 OK" << endl;
    return 0;
}
